ts/nni_manager/common/command_channel/websocket.ts

```diff
@@ -0,0 +1,207 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+/**
+ *  WebSocket command channel server.
+ *
+ *  By default the server maintains a heartbeat for each client.
+ *  When a client loses heartbeat for 20 seconds, the connection will be closed.
+ *
+ *  The constructor provides an optional parameter `robustness` to control this behavior.
+ *  The parameter specifies how many consecutive missing heartbeats (at 5s interval) will kill a connection.
+ *  Setting the robustness to 0 will drop dead connections in 5 seconds,
+ *  and settig it to `Infinity` will effectively disable the heartbeat.
+ *
+ *  Note that the WebSocket connection is not reliable, even for localhost loopback connection.
+ *  The client should recreate a connection when any error occurs.
+ *
+ *  To keep consistency with other channel types, each URL only serves one connection.
+ *  When the client creates a new connection to the same URL, the previous connection will be closed by server.
+ **/
+
+import { EventEmitter } from 'events';
+
+import type { Request } from 'express';
+import type { WebSocket } from 'ws';
+
+import { Deferred } from 'common/deferred';
+import globals from 'common/globals';
+import { Logger, getLogger } from 'common/log';
+import type { Command, CommandChannelServer } from './interface';
+
+let heartbeatInterval: number = 5000;
+
+type ReceiveCallback = (channelId: string, command: Command) => void;
+
+export class WebSocketChannelServer implements CommandChannelServer {
```

Comment:
do you have related code that uses `WebSocketChannelServer` as client in training service?